datafusion/physical-plan/src/aggregates/group_values/group_value_row.rs

```diff
@@ -0,0 +1,446 @@
+            let first_n = self.group_values.drain(0..n).collect::<Vec<_>>();
+            let first_n_nulls = self.nulls.drain(0..n).collect::<Vec<_>>();
+            Arc::new(PrimitiveArray::<T>::new(
+                ScalarBuffer::from(first_n),
+                Some(NullBuffer::from(first_n_nulls)),
+            ))
+        } else {
+            let first_n = self.group_values.drain(0..n).collect::<Vec<_>>();
+            self.nulls.truncate(self.nulls.len() - n);
+            Arc::new(PrimitiveArray::<T>::new(ScalarBuffer::from(first_n), None))
+        }
+    }
+}
+
+pub struct ByteGroupValueBuilder<O>
+where
+    O: OffsetSizeTrait,
+{
+    output_type: OutputType,
+    buffer: BufferBuilder<u8>,
+    /// Offsets into `buffer` for each distinct  value. These offsets as used
+    /// directly to create the final `GenericBinaryArray`. The `i`th string is
+    /// stored in the range `offsets[i]..offsets[i+1]` in `buffer`. Null values
+    /// are stored as a zero length string.
+    offsets: Vec<O>,
+    /// Null indexes in offsets, if `i` is in nulls, `offsets[i]` should be equals to `offsets[i+1]`
+    nulls: Vec<usize>,
+}
+
+impl<O> ByteGroupValueBuilder<O>
+where
+    O: OffsetSizeTrait,
+{
+    pub fn new(output_type: OutputType) -> Self {
+        Self {
+            output_type,
+            buffer: BufferBuilder::new(INITIAL_BUFFER_CAPACITY),
+            offsets: vec![O::default()],
+            nulls: vec![],
+        }
+    }
+
+    fn append_val_inner<B>(&mut self, array: &ArrayRef, row: usize)
+    where
+        B: ByteArrayType,
+    {
+        let arr = array.as_bytes::<B>();
+        if arr.is_null(row) {
+            self.nulls.push(self.len());
+            // nulls need a zero length in the offset buffer
+            let offset = self.buffer.len();
+
+            self.offsets.push(O::usize_as(offset));
+            return;
+        }
+
+        let value: &[u8] = arr.value(row).as_ref();
+        self.buffer.append_slice(value);
+        self.offsets.push(O::usize_as(self.buffer.len()));
+    }
+
+    fn equal_to_inner<B>(&self, lhs_row: usize, array: &ArrayRef, rhs_row: usize) -> bool
+    where
+        B: ByteArrayType,
+    {
+        // Handle nulls
+        let is_lhs_null = self.nulls.iter().any(|null_idx| *null_idx == lhs_row);
+        let arr = array.as_bytes::<B>();
+        if is_lhs_null {
+            return arr.is_null(rhs_row);
+        } else if arr.is_null(rhs_row) {
+            return false;
+        }
+
+        let arr = array.as_bytes::<B>();
+        let rhs_elem: &[u8] = arr.value(rhs_row).as_ref();
+        let rhs_elem_len = arr.value_length(rhs_row).as_usize();
+        debug_assert_eq!(rhs_elem_len, rhs_elem.len());
+        let l = self.offsets[lhs_row].as_usize();
+        let r = self.offsets[lhs_row + 1].as_usize();
+        let existing_elem = unsafe { self.buffer.as_slice().get_unchecked(l..r) };
+        existing_elem.len() == rhs_elem.len() && rhs_elem == existing_elem
```

Comment:
Do we really need to check the length manually? Will that not just be a part of the `==` for slices. (https://doc.rust-lang.org/1.81.0/src/core/slice/cmp.rs.html#59)
```suggestion
       rhs_elem == existing_elem
```